ompi/communicator/comm_ft_detector.c

```diff
@@ -0,0 +1,742 @@
+        ret = opal_thread_start(&fd_thread);
+        if( OPAL_SUCCESS != ret ) goto cleanup;
+        while( 0 == fd_thread_active ); /* wait for the fd thread initialization */
+        if( 0 > fd_thread_active ) goto cleanup;
+    }
+
+    return OMPI_SUCCESS;
+
+  cleanup:
+    ompi_comm_failure_detector_finalize();
+    return ret;
+}
+
+int ompi_comm_failure_detector_start(void) {
+    /* fd not wanted */
+    if( -1 == comm_heartbeat_recv_cb_type ) return OMPI_SUCCESS;
+
+    if( comm_detector_use_thread ) {
+        OPAL_THREAD_ADD_FETCH32(&fd_thread_active, 1);
+        return OMPI_SUCCESS;
+    }
+
+    /* setting up the default detector on comm_world */
+    return ompi_comm_start_detector(&ompi_mpi_comm_world.comm);
+}
+
+#define FD_LOCAL_PROCS 1
+
+int ompi_comm_failure_detector_finalize(void) {
+    int observing;
+    comm_detector_t* detector = &comm_world_detector;
+
+    /* Tell our observer that we won't put anymore */
+    if( MPI_PROC_NULL != detector->hb_observer ) {
+        detector->hb_rdma_rank = detector->hb_observer;
+        fd_heartbeat_send(detector);
+        detector->hb_period = INFINITY;
+        detector->hb_observer = MPI_PROC_NULL;
+        opal_atomic_mb();
+    }
+    /* wait until the observed process confirms he is not putting in our
+     * memory (or everybody else is dead) */
+    while( MPI_PROC_NULL != (observing = detector->hb_observing) ) {
+        ompi_proc_t* proc = ompi_comm_peer_lookup(detector->comm, observing);
+        assert( NULL != proc );
+#if !FD_LOCAL_PROCS
+        if( OPAL_PROC_ON_LOCAL_NODE(proc->super.proc_flags) ) {
+            break;
+        }
+#endif
+        while( observing == detector->hb_observing ) {
+            /* If observed process changed, recheck if local*/
+            if( !(0 < fd_thread_active) )
+            {
+                opal_progress();
+            }
+        }
+    }
+
+    if( 0 < fd_thread_active ) {
+        void* tret;
+        /* this is not a race condition. Accesses are serialized, we use the
+         * atomic for the mfence part of it. */
+        OPAL_THREAD_ADD_FETCH32(&fd_thread_active, -fd_thread_active);
+        opal_event_base_loopbreak(fd_event_base);
+        opal_thread_join(&fd_thread, &tret);
+    }
+
+    if( NULL != detector->fd_event ) {
+        opal_event_del(detector->fd_event);
+        opal_event_free(detector->fd_event);
+        detector->fd_event = NULL;
+    }
+    if( opal_sync_event_base != fd_event_base ) opal_event_base_free(fd_event_base);
+
+    /* Remove rdma registrations, if any */
+    if( NULL != detector->hb_rdma_rank_lreg ) {
+        mca_bml_base_deregister_mem(detector->hb_rdma_bml_btl_observer, detector->hb_rdma_rank_lreg);
+    }
+    if( NULL != detector->hb_rdma_flag_lreg ) {
+        mca_bml_base_deregister_mem(detector->hb_rdma_bml_btl_observing, detector->hb_rdma_flag_lreg);
+    }
+
+#if 0 /* not doing this permits an optimization in finalize and saves an extra barrier */
```

Comment:
hard-coded conditional: shouldn't the PR show only the choice that works today, and alternative code get pushed to a different branch?